modules/plugin/shapefile/src/test/java/org/geotools/data/shapefile/IndexManagerTest.java

```diff
@@ -0,0 +1,228 @@
+        IndexManager indexManager = new IndexManager(shpFiles, mockDataStore);
+        if (!shpFiles.exists(QIX)) {
+            assertTrue(indexManager.createSpatialIndex(true));
+        }
+
+        File shpFile = null;
+        File indexFile = null;
+        try {
+            shpFile = shpFiles.acquireReadFile(SHP, indexManager.writer);
+            indexFile = shpFiles.acquireReadFile(QIX, indexManager.writer);
+            assertTrue(indexFile.exists());
+        } finally {
+            shpFiles.unlockRead(indexFile, indexManager.writer);
+            shpFiles.unlockRead(shpFile, indexManager.writer);
+        }
+
+        // make the file stale (i.e. older than the .shp)
+        indexFile.setLastModified(shpFile.lastModified() - 100_000);
+        final long lastModified = indexFile.lastModified();
+
+        final boolean force = false;
+        assertTrue("stale index should force rebuild", indexManager.createSpatialIndex(force));
+        assertTrue(shpFiles.exists(QIX));
+
+        assertTrue(indexFile.lastModified() > lastModified);
+
+        assertFalse("index is not stale anymore", indexManager.createSpatialIndex(force));
+    }
+
+    @Test(timeout = 10_000) // timeout at 10s to fail in case of deadlock
+    public void testCreateSpatialIndex_Concurrency_runs_once() throws Exception {
+        deleteQixFile();
+        IndexManager indexManager = new IndexManager(shpFiles, mockDataStore);
+
+        final int threadCount = 8;
+        final int taskCount = 16;
+        final boolean force = false;
+        List<Boolean> builds =
+                buildSpatialIndexConcurrently(indexManager, threadCount, taskCount, force);
+
+        long buildCount = builds.stream().filter(Boolean::booleanValue).count();
+
+        String msg =
+                String.format(
+                        "Expected only 1 true result from %d concurrent createSpatialIndex calls",
+                        taskCount);
+        assertEquals(msg, 1, buildCount);
+    }
+
+    @Test(timeout = 10_000) // timeout at 10s to fail in case of deadlock
+    public void testCreateSpatialIndex_Concurrency_with_force_run_sequentially() throws Exception {
+        deleteQixFile();
+        final AtomicInteger concurrentBuilds = new AtomicInteger();
+        final AtomicInteger maxConcurrentBuilds = new AtomicInteger();
+        IndexManager indexManager =
+                new IndexManager(shpFiles, mockDataStore) {
+                    @Override
+                    protected void doCreateSpatialIndex() throws Exception {
+                        int concurrency = concurrentBuilds.incrementAndGet();
+                        super.doCreateSpatialIndex();
+                        concurrentBuilds.decrementAndGet();
+                        maxConcurrentBuilds.set(Math.max(maxConcurrentBuilds.get(), concurrency));
+                    }
+                };
+
+        final int threadCount = 8;
```

Comment:
Try to relate the number of threads to the actual CPUs, on a build server or PR check we're unlikely to have that many. 
Maybe something like this:

```suggestion
        final int threadCount = Math.min(8, Runtime.getRuntime().availableProcessors());
```

Reply:
https://github.com/geotools/geotools/pull/3527/files#diff-cca0c8a31b2bd3b8cf2a2473aae2301caf3f80def759a5d36a2d82b3cc4e319bR173-R174